lib/timers.js

```diff
@@ -54,99 +128,132 @@ exports.active = function(item) {
 };
 
 function listOnTimeout() {
-  var msecs = this.msecs;
-  var list = this;
+  var list = this._list;
+  var msecs = list.msecs;
 
   debug('timeout callback %d', msecs);
 
-  var now = Timer.now();
+  var now = TimerWrap.now();
   debug('now: %s', now);
 
-  var diff, first, threw;
-  while (first = L.peek(list)) {
-    diff = now - first._idleStart;
+  var diff, timer, threw;
+  while (timer = L.peek(list)) {
+    diff = now - timer._idleStart;
+
+    // Check if this loop iteration is too early for the next timer.
+    // This happens if there are more timers scheduled for later in the list.
     if (diff < msecs) {
-      list.start(msecs - diff, 0);
+      this.start(msecs - diff, 0);
       debug('%d list wait because diff is %d', msecs, diff);
       return;
-    } else {
-      L.remove(first);
-      assert(first !== L.peek(list));
+    }
+
+    // The actual logic for when a timeout happens.
 
-      if (!first._onTimeout) continue;
+    L.remove(timer);
+    assert(timer !== L.peek(list));
 
-      // v0.4 compatibility: if the timer callback throws and the
+    if (!timer._onTimeout) continue;
+
+    var domain = timer.domain;
+    if (domain) {
+
+      // If the timer callback throws and the
       // domain or uncaughtException handler ignore the exception,
       // other timers that expire on this tick should still run.
       //
-      // https://github.com/joyent/node/issues/2631
-      var domain = first.domain;
-      if (domain && domain._disposed)
+      // https://github.com/nodejs/node-v0.x-archive/issues/2631
+      if (domain._disposed)
         continue;
 
-      try {
-        if (domain)
-          domain.enter();
-        threw = true;
-        first._called = true;
-        first._onTimeout();
-        if (domain)
-          domain.exit();
-        threw = false;
-      } finally {
-        if (threw) {
-          // We need to continue processing after domain error handling
-          // is complete, but not by using whatever domain was left over
-          // when the timeout threw its exception.
-          var oldDomain = process.domain;
-          process.domain = null;
-          process.nextTick(listOnTimeoutNT, list);
-          process.domain = oldDomain;
-        }
-      }
+      domain.enter();
     }
+
+    tryOnTimeout(timer, list);
+
+    if (domain)
+      domain.exit();
   }
 
+  // If `L.peek(list)` returned nothing, the list was either empty or we have
+  // called all of the timer timeouts.
+  // As such, we can remove the list and clean up the TimerWrap C++ handle.
   debug('%d list empty', msecs);
   assert(L.isEmpty(list));
-  list.close();
-  delete lists[msecs];
+  this.close();
+  if (list._unrefed === true) {
+    delete unrefedLists[msecs];
+  } else {
+    delete refedLists[msecs];
```

Comment:
Perhaps we could get the best of both worlds by tombstone-ing using `null`, keeping a count of tombstones, and copying when the number gets above a threshold to avoid reverting to slow-mode (`512`?) This might not be worthwhile given the numbers, though, and could be tested in a subsequent PR.